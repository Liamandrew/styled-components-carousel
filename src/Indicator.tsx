import React from 'react';
import styled from 'styled-components';

type Props = {
    items: number;
    slidesToShow: number;
    infinite?: boolean;
    active: number;
    onClick: (index: number) => void;
};

export const IndicatorWrapper = styled.div`
    display: flex;
    flex-direction: row;
    padding: 24px 12px;
`;

export const Indicator = styled.button<{ highlighted: boolean }>`
    width: 10px;
    height: 10px;
    border-radius: 50%;
    padding: 0;
    border: 0;
    outline: 0;
    margin-left: 14px;
    margin-right: 14px;
    background-color: ${props => (props.highlighted ? '#444444' : '#DDDDDD')};
    transition: transform 0.1s;
    transform: scale(${props => (props.highlighted ? '1.1' : '1')});

    &:hover {
        transition: transform 0.1s;
        transform: scale(1.1);
    }
`;

const getNumberOfIndicators = (items: number, slidesToShow: number) => {
    if (slidesToShow >= items) {
        return 1;
    }

    return items;
};

const Component: React.FC<Props> = ({ items, slidesToShow, infinite, active, onClick }) => {
    return (
        <IndicatorWrapper>
            {[...Array(getNumberOfIndicators(items, slidesToShow))].map((e, i) => (
                <Indicator
                    key={`carousel-indicator-${i}`}
                    data-testid={`carousel-indicator-${i}`}
                    highlighted={i === active % items}
                    onClick={() => onClick(i)}
                />
            ))}
        </IndicatorWrapper>
    );
};

export default Component;
