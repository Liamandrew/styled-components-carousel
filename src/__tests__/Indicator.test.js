import React from 'react';
import { render, fireEvent, getByTestId } from '@testing-library/react';
import Indicator from '../Indicator';

describe('Indicator', () => {
    const onClick = jest.fn();

    beforeEach(() => {
        jest.clearAllMocks();
    });

    it('should match the snapshot', () => {
        const { container } = render(<Indicator items={3} slidesToShow={2} infinite active={2} onClick={onClick} />);

        expect(container).toMatchSnapshot();
    });

    it('should match the snapshot when slides to show is greater than items', () => {
        const { container } = render(<Indicator items={3} slidesToShow={4} infinite active={2} onClick={onClick} />);

        expect(container).toMatchSnapshot();
    });

    it('should call onClick with the correct index', () => {
        const { container } = render(<Indicator items={3} slidesToShow={2} infinite active={2} onClick={onClick} />);

        fireEvent.click(getByTestId(container, 'carousel-indicator-1'));

        expect(onClick).toHaveBeenCalledWith(1);
    });
});
