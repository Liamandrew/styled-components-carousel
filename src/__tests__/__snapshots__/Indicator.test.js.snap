// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Indicator should match the snapshot 1`] = `
<div>
  <div
    class="sc-AxjAm hxQek"
  >
    <button
      class="sc-AxirZ kNLnTd"
      data-testid="carousel-indicator-0"
    />
    <button
      class="sc-AxirZ kNLnTd"
      data-testid="carousel-indicator-1"
    />
    <button
      class="sc-AxirZ dBFIEW"
      data-testid="carousel-indicator-2"
    />
  </div>
</div>
`;

exports[`Indicator should match the snapshot when slides to show is greater than items 1`] = `
<div>
  <div
    class="sc-AxjAm hxQek"
  >
    <button
      class="sc-AxirZ kNLnTd"
      data-testid="carousel-indicator-0"
    />
    <button
      class="sc-AxirZ kNLnTd"
      data-testid="carousel-indicator-1"
    />
    <button
      class="sc-AxirZ dBFIEW"
      data-testid="carousel-indicator-2"
    />
  </div>
</div>
`;
